src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -994,30 +1029,38 @@ internal static string GetAsyncStateMachineDescription(IAsyncStateMachine stateM
         /// <param name="targetContext">The target context on which to propagate the exception.  Null to use the ThreadPool.</param>
         internal static void ThrowAsync(Exception exception, SynchronizationContext targetContext)
         {
-            if (exception == null) throw new ArgumentNullException(nameof(exception));
+            // Capture the exception into an ExceptionDispatchInfo so that its 
+            // stack trace and Watson bucket info will be preserved
+            var edi = ExceptionDispatchInfo.Capture(exception);
 
             // If the user supplied a SynchronizationContext...
             if (targetContext != null)
             {
                 try
                 {
-                    // Capture the exception into an ExceptionDispatchInfo so that its 
-                    // stack trace and Watson bucket info will be preserved
-                    var edi = ExceptionDispatchInfo.Capture(exception);
-
                     // Post the throwing of the exception to that context, and return.
                     targetContext.Post(state => ((ExceptionDispatchInfo)state).Throw(), edi);
                     return;
                 }
                 catch (Exception postException)
                 {
-                    // If something goes horribly wrong in the Post, we'll treat this a *both* exceptions
-                    // going unhandled.
-                    RuntimeAugments.ReportUnhandledException(new AggregateException(exception, postException));
+                    // If something goes horribly wrong in the Post, we'll 
+                    // propagate both exceptions on the ThreadPool
+                    edi = ExceptionDispatchInfo.Capture(new AggregateException(exception, postException));
                 }
             }
 
-            RuntimeAugments.ReportUnhandledException(exception);
+#if CORERT
+            RuntimeAugments.ReportUnhandledException(edi.SourceException);
+#else
+            // If we have the new error reporting APIs, report this error.  Otherwise, Propagate the exception(s) on the ThreadPool
+#if FEATURE_COMINTEROP
+            if (!WindowsRuntimeMarshal.ReportUnhandledError(edi.SourceException))
+#endif // FEATURE_COMINTEROP
+            {
+                ThreadPool.QueueUserWorkItem(state => ((ExceptionDispatchInfo)state).Throw(), edi);
+            }
+#endif
```

Comment:
Not sure how to proceed on this one `#if CORERT`.

Reply:
Yes, having it under ifdef is best option, for now at least. Maybe tweak the formatting a little bit to make it more readable:

```
#if CORERT
    RuntimeAugments.ReportUnhandledException(edi.SourceException);
#else

#if FEATURE_COMINTEROP
    // If we have the new error reporting APIs, report this error.
    if (WindowsRuntimeMarshal.ReportUnhandledError(edi.SourceException))
        return;
#endif // FEATURE_COMINTEROP

    // Propagate the exception(s) on the ThreadPool
    ThreadPool.QueueUserWorkItem(state => ((ExceptionDispatchInfo)state).Throw(), edi);

#endif // CORERT
```

Reply:
Applied your suggestion verbatim, thanks!